### PR TITLE
Show login failure message.

### PR DESCRIPTION
**Pull request by grokys**
Initial fix for #351. Shows the error message when connection fails for .com and enterprise.

Needs a design pass! Also not sure if binding to the containing `ErrorMessageDisplay.Message` property is elegant or a hack.

image

image

### CODE REVIEW

**Comment by donokuda** [comment]
> Needs a design pass!

I'll open a separate PR with a design pass 🎨

**Comment by shana** [comment]
There's already code that handles showing detailed error messages being used elsewhere, I honestly don't know why we're not doing that (slipped under the radar, I'm sure)

The repo creation viewmodel uses UserErrors.Throw to throw exceptions that then get caught by the `<uirx:UserErrorMessages />` xaml object. 

Throwing:
https://github.com/github/VisualStudio/blob/master/src/GitHub.App/ViewModels/RepositoryCreationViewModel.cs#L296

Showing:
https://github.com/github/VisualStudio/blob/master/src/GitHub.VisualStudio/UI/Views/Controls/RepositoryCreationControl.xaml#L243

(UserErrorMessage registers a handler for the Throw method in https://github.com/github/VisualStudio/blob/master/src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs#L111-L111, fyi)

So technically what we want to do is have a `<uirx:UserErrorMessages>` block and then use `UserError.Throw` to show the messages (optionally changing the wording on them if we want)

**Comment by donokuda** [comment]
@shana Should I hold off on polish until your suggestions are brought in?

**Comment by grokys** [comment]
@shana ah! Yes I'd forgotten about that control. Will change to use that.

@donokuda yep, hold off until I make the changes @shana suggested.

**Comment by shana** [comment]
@donokuda It might not need polish at all, we should already have all the right stuff put in place for this.

**Comment by donokuda** [comment]
aww yeahhh! Feel free to 🔔 me if you do need any polish! ✨

**Comment by grokys** [comment]
Now updated to use `UserError.Throw` throughout the login viewmodel. The required adding support for recovery options in `UserErrorMessage`. Here's what it looks like now:

Bad username/pass .com:

image

Bad username/pass enterprise:

image

Other error dotcom:

image

File: src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs [hunk 1]
```diff
@@ -46,7 +46,7 @@ public UserErrorMessages()
             };
         }
 
-        public static readonly DependencyProperty IconMarginProperty = DependencyProperty.Register("IconMargin", typeof(Thickness), typeof(UserErrorMessages), new PropertyMetadata(new Thickness(0,10,7,0)));
+        public static readonly DependencyProperty IconMarginProperty = DependencyProperty.Register("IconMargin", typeof(Thickness), typeof(UserErrorMessages), new PropertyMetadata(new Thickness(0,0,8,0)));
```
Comment (shana) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: This is going to affect all the other dialogs that don't set this property explicitly. Are we sure it's not going to break existing error messages UI?

File: src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs [hunk 2]
```diff
@@ -106,18 +106,14 @@ public UserError UserError
             Justification = "We're registering a handler for a type so this is appropriate.")]
         public IDisposable RegisterHandler<TUserError>(IObservable<bool> clearWhen) where TUserError : UserError
         {
-            if (IsVisible)
+            return UserError.RegisterHandler<TUserError>(userError =>
```
Comment (shana) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: What's the reason for this change?

File: src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs [hunk 3]
```diff
@@ -22,6 +24,8 @@ public class GenericLoginControl : SimpleViewUserControl<ILoginControlViewModel,
     [PartCreationPolicy(CreationPolicy.NonShared)]
     public partial class LoginControl : GenericLoginControl
     {
+        private IDisposable errorHandler;
```
Comment (shana) [inline src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs]: `private` :gun:

File: src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs [hunk 4]
```diff
@@ -39,7 +43,10 @@ public LoginControl()
                         NotifyDone();
                     }
                 });
+
+                d(Disposable.Create(() => errorHandler.Dispose()));
```
Comment (shana) [inline src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs]: Can `errorHandler` be null here? It's only set in the selection changed event below, and you check for nullness there, is that event guaranteed to be invoked always?

File: src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs [hunk 5]
```diff
@@ -112,5 +107,25 @@ void SetupSelectedAndVisibleTabBindings(Action<IDisposable> d)
                 .Where(x => x == true)
                 .BindTo(this, v => v.enterpriseTab.IsSelected));
         }
+
+        private void hostTabControl_SelectionChanged(object sender, System.Windows.Controls.SelectionChangedEventArgs e)
```
Comment (shana) [inline src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs]: `private` :gun:

File: src/GitHub.App/ViewModels/LoginTabViewModel.cs [hunk 6]
```diff
@@ -18,6 +19,7 @@
 
 namespace GitHub.ViewModels
 {
+    [SuppressMessage("Microsoft.Design", "CA1001:TypesThatOwnDisposableFieldsShouldBeDisposable")]
```
Comment (shana) [inline src/GitHub.App/ViewModels/LoginTabViewModel.cs]: Is this still needed?

**Comment by shana** [comment]
Looking good, left a few comments sprinkled about.  🏈 ➡️

File: src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs (hunk 2 above)
Comment (grokys) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: The handler in the login dialog wasn't getting registered because the `UserErrorMessages` control wasn't visible at the time of the call. I saw that check and it looked wrong: if someone wants to not register a handler when a control isn't visible they should do it explicitly IMO. Doing nothing in a situation like this just seems confusing to me...

File: src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs (hunk 4 above)
Comment (grokys) [inline src/GitHub.VisualStudio/UI/Views/Controls/LoginControl.xaml.cs]: I don't think so - a tab will be selected when the UI is built, causing the selection changed event to fire. Looking at the WPF call stack it looks like this will always happen.

File: src/GitHub.App/ViewModels/LoginTabViewModel.cs (hunk 6 above)
Comment (grokys) [inline src/GitHub.App/ViewModels/LoginTabViewModel.cs]: Yeah, because it creates a `RecoveryCommand` (which implements `IDisposable`) and assigns it to `NavigateForgotPassword`. I don't think `RecoveryCommand` actually needs to be disposed though - it appears that the only purpose of it is to set its `CanExecute` state.

Should I dispose the `RecoveryCommand` anyway? Or just add a justification?

File: src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs (hunk 1 above)
Comment (grokys) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: I managed to get an error message to show in ``RepositoryCreationControl` and it looks fine I think. TBH not sure why it wasn't off-center there.

Comment (grokys) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: Just tried `RepositoryCreationControl` on master and the error doesn't even seem to appear, which is probably why nobody's noticed it's off-center :)

File: src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs (hunk 2 above)
Comment (shana) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: heh, yeah, makes sense (or rather, it doesn't...) 👍

File: src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs (hunk 1 above)
Comment (shana) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: Hmmm, I wonder if the error there is meant to appear... we do have some errors that show up inline.

Comment (shana) [inline src/GitHub.UI.Reactive/Controls/Validation/UserErrorMessages.cs]: Ah, it's the creation repo failure message. I guess we never showed that then? :P

File: src/GitHub.App/ViewModels/LoginTabViewModel.cs (hunk 6 above)
Comment (shana) [inline src/GitHub.App/ViewModels/LoginTabViewModel.cs]: `RecoveryCommand` leaks a disposable for the `canExecute`, but it should be disposed automatically when the viewmodel gets disposed of (which should happen when the windows is closed), so it should be fine.
